src/rust/engine/process_execution/src/speculate.rs

```diff
@@ -0,0 +1,184 @@
+        }
+      }
+    ).to_boxed()
+  }
+}
+
+#[cfg(test)]
+mod tests {
+  use bytes::Bytes;
+  use futures::future::Future;
+  use hashing::{EMPTY_DIGEST};
+  use boxfuture::{BoxFuture, Boxable};
+  use std::time::{Duration, Instant};
+  use tokio_timer::Delay;
+  use tokio;
+  use testutil::owned_string_vec;
+  use std::collections::{BTreeMap, BTreeSet};
+  use workunit_store::WorkUnitStore;
+
+  use super::{
+    SpeculatingCommandRunner,
+    CommandRunner,
+    ExecuteProcessRequest,
+    FallibleExecuteProcessResult,
+  };
+
+
+  #[test]
+  fn test_no_speculation() {
+    let result = run_speculation_test(0, 0, 100, false, false,);
+    assert_eq![result.unwrap().stdout, Bytes::from("m1")]
+  }
+
+  #[test]
+  fn test_speculate() {
+    let result = run_speculation_test(100, 0, 10, false, false);
+    assert_eq![result.unwrap().stdout, Bytes::from("m2")]
+  }
+
+  #[test]
+  fn first_req_slow_success() {
+    let result = run_speculation_test(15, 10, 10, false, false);
+    assert_eq![result.unwrap().stdout, Bytes::from("m1")]
+  }
+
+  #[test]
+  fn first_req_slow_fail() {
+    let result = run_speculation_test(100, 0, 10, true, false);
+    assert_eq![result.unwrap().stdout, Bytes::from("m2")]
+  }
+
+  #[test]
+  fn first_req_fast_fail() {
+    let result = run_speculation_test(15, 10, 10, true, false);
+    assert_eq![result.unwrap_err(), Bytes::from("m1")]
+  }
+
+  #[test]
+  fn second_req_fast_fail() {
+    let result = run_speculation_test(100, 0, 10, true, true);
+    assert_eq![result.unwrap_err(), Bytes::from("m2")]
+  }
+
+  fn run_speculation_test(
+    r1_latency_ms: u64,
+    r2_latency_ms: u64,
+    speculation_delay_ms: u64,
+    r1_is_err: bool,
+    r2_is_err: bool,
+  ) -> Result<FallibleExecuteProcessResult, String> {
+    let mut runtime = tokio::runtime::Runtime::new().unwrap();
+    let execute_request = echo_foo_request();
+    let msg1: String = "m1".into();
+    let msg2: String = "m2".into();
+    let workunit_store = WorkUnitStore::new();
+    let runner = SpeculatingCommandRunner::new(
+      Box::new(make_delayed_command_runner(msg1.clone(), r1_latency_ms, r1_is_err)),
+      Box::new(make_delayed_command_runner(msg2.clone(), r2_latency_ms, r2_is_err)),
+      Duration::from_millis(speculation_delay_ms),
+    );
+    runtime.block_on(runner.run(execute_request, workunit_store))
+  }
+
+  fn make_delayed_command_runner(msg: String, delay: u64, is_err: bool) -> DelayedCommandRunner {
+    let mut result;
+    if is_err {
+      result = Err(msg.into());
+    } else {
+      result = Ok(FallibleExecuteProcessResult {
+        stdout: msg.into(),
+        stderr: "".into(),
+        exit_code: 0,
+        output_directory: EMPTY_DIGEST,
+        execution_attempts: vec![],
+      })
+    }
+    DelayedCommandRunner::new(Duration::from_millis(delay), result)
+  }
+
+  fn echo_foo_request() -> ExecuteProcessRequest {
```

Comment:
FWIW you should be able to:
`use crate::remote::tests::echo_foo_request;`
which may need you to put a `pub` before `fn echo_foo_request` in the other file
and things should Just Work :)